platform/ios/app/MBXFrameTimeGraphView.m

```diff
@@ -0,0 +1,127 @@
+#import "MBXFrameTimeGraphView.h"
+
+const CGFloat MBXFrameTimeExaggeration = 4.f * 1000.f;
+const CGFloat MBXFrameTimeBarWidth = 4.f;
+
+@interface MBXFrameTimeGraphView ()
+
+@property (nonatomic) CAScrollLayer *scrollLayer;
+@property (nonatomic) CAShapeLayer *thresholdLayer;
+@property (nonatomic) CGFloat currentX;
+@property (nonatomic) NSMutableArray<CAShapeLayer *> *barLayers;
+
+@end
+
+@implementation MBXFrameTimeGraphView
+
+- (instancetype)init {
+    if (self = [super init]) {
+        [self commonInit];
+    }
+    return self;
+}
+
+- (instancetype)initWithCoder:(NSCoder *)aDecoder {
+    if (self = [super initWithCoder:aDecoder]) {
+        [self commonInit];
+    }
+    return self;
+}
+
+- (void)commonInit {
+    self.userInteractionEnabled = NO;
+    self.layer.opacity = 0.9f;
+
+    self.scrollLayer = [CAScrollLayer layer];
+    self.scrollLayer.scrollMode = kCAScrollHorizontally;
+    self.scrollLayer.masksToBounds = YES;
+    [self.layer addSublayer:self.scrollLayer];
+
+    self.thresholdLayer = [CAShapeLayer layer];
+    self.thresholdLayer.fillColor = [UIColor darkGrayColor].CGColor;
+    [self.layer addSublayer:self.thresholdLayer];
+
+    self.barLayers = [NSMutableArray array];
+}
+
+- (void)layoutSubviews {
+    [super layoutSubviews];
+
+    if (!CGRectEqualToRect(self.scrollLayer.frame, self.bounds)) {
+        self.scrollLayer.frame = self.bounds;
+
+        CGRect thresholdLineRect = CGRectMake(0, self.frame.size.height - [self renderDurationTargetMilliseconds], self.frame.size.width, 1);
+        self.thresholdLayer.path = CGPathCreateWithRect(thresholdLineRect, nil);
+    }
+}
+
+- (void)updatePathWithFrameDuration:(CFTimeInterval)frameDuration {
+    [CATransaction begin];
+    [CATransaction setDisableActions:YES];
+
+    self.currentX += MBXFrameTimeBarWidth;
+
+    CAShapeLayer *bar = [self barWithFrameDuration:frameDuration];
+    bar.position = CGPointMake(self.currentX, self.frame.size.height);
+
+    [self.scrollLayer addSublayer:bar];
+    [self.barLayers addObject:bar];
+
+    [self.scrollLayer scrollToPoint:CGPointMake(self.currentX - self.frame.size.width, 0)];
+
+    [self removeStaleBars];
+
+    [CATransaction commit];
+}
+
+- (CGFloat)renderDurationTargetMilliseconds {
+    CGFloat target = (1.0 / UIScreen.mainScreen.maximumFramesPerSecond) * MBXFrameTimeExaggeration;
+    return [self roundedFloat:target];
+}
+
+- (CGFloat)roundedFloat:(CGFloat)f {
+#if TARGET_OS_IPHONE || TARGET_OS_SIMULATOR
+    CGFloat scaleFactor = [UIScreen mainScreen].nativeScale;
+#elif TARGET_OS_MAC
+    CGFloat scaleFactor = [NSScreen mainScreen].backingScaleFactor;
+#endif
+    return round(f * scaleFactor) / scaleFactor;
+}
+
+- (CAShapeLayer *)barWithFrameDuration:(CFTimeInterval)frameDuration {
+    CAShapeLayer *bar = [CAShapeLayer layer];
```

Comment:
How expensive is this operation? We could consider a reuse pool.

Reply:
This object is apparently fairly cheap to initialize — 0.0% in profiling:

<img width="888" alt="screen shot 2018-11-27 at 2 33 51 pm" src="https://user-images.githubusercontent.com/1198851/49106609-a3d94c80-f251-11e8-9653-f0b82ca34b07.png">